README.md
Description
-----------

[![Build Status](https://travis-ci.org/ActiveTriples/ActiveTriples.png?branch=develop)](https://travis-ci.org/ActiveTriples/ActiveTriples)
[![Coverage Status](https://coveralls.io/repos/ActiveTriples/ActiveTriples/badge.svg?branch=develop)](https://coveralls.io/r/ActiveTriples/ActiveTriples?branch=develop)
[![Gem Version](https://badge.fury.io/rb/active-triples.svg)](http://badge.fury.io/rb/active-triples)

An ActiveModel-like interface for RDF data. Models graphs as RDFSources with property/attribute configuration, accessors, and other methods to support Linked Data in a Ruby/Rails enviornment. See [RDF Concepts and Abstract Syntax](http://www.w3.org/TR/2014/REC-rdf11-concepts-20140225/#change-over-time) for an informal definition of an RDF Source.

This library was extracted from work on [ActiveFedora](https://github.com/projecthydra/active_fedora). It is closely related to (and borrows some syntax from) [Spira](https://github.com/ruby-rdf/spira), but does some important things differently.

Installation
------------

Add `gem "active-triples"` to your Gemfile and run `bundle`.

Or install manually with `gem install active-triples`.

Defining RDFSource Models
-------------------------

The core module of `ActiveTriples` is `ActiveTriples::RDFSource`. You can use this module as a mixin to create ActiveModel-like classes that represent an RDF resource as a stateful entity represented by an `RDF::Graph`. `RDFSource` implements the `RDF::Resource` interface, as well as `RDF::Queryable`, `RDF::Enumerable`, and `RDF::Mutable`. This means you can manipulate them by adding or deleting statements, query, serialize, and load arbitrary RDF.


```ruby
require 'rdf/vocab'

class Thing
  include  ActiveTriples::RDFSource
  
  configure type: RDF::OWL.Thing, base_uri: 'http://example.org/things#'
  
  property :title,       predicate: RDF::Vocab::DC.title
  property :description, predicate: RDF::Vocab::DC.description
end

obj             = Thing.new('123')
obj.title       = 'Resource'
obj.description = 'A resource.'

obj.dump :ntriples # => "<http://example.org/things#123> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .\n<http://example.org/things#123> <http://purl.org/dc/terms/title> \"Resource\" .\n<http://example.org/things#123> <http://purl.org/dc/terms/description> \"A resource.\" .\n"
```

URI and bnode values are built out as generic Resources when accessed. A more specific model class can be configured on individual properties.

```ruby
Thing.property :creator, predicate: RDF::Vocab::DC.creator, class_name: 'Person'

class Person
  include  ActiveTriples::RDFSource
  
  configure type:     RDF::Vocab::FOAF.Person,
            base_uri: 'http://example.org/people#'
            
  property :name, predicate: RDF::Vocab::FOAF.name
end

obj_2         = Thing.new('2')
obj_2.creator = Person.new

obj_2.creator
# => [#<Person:0x3fbe84ac9234(default)>]

obj_2.creator.first.name = 'Herman Melville'

obj_2.dump :ntriples # => "_:g47361345336040 <http://xmlns.com/foaf/0.1/name> \"Herman Melville\" .\n_:g47361345336040 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> .\n<http://example.org/things#2> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .\n<http://example.org/things#2> <http://purl.org/dc/terms/creator> _:g47361345336040 .\n"
```

Open Model
-----------

An RDFSource lets you handle data as a graph, independent of whether it is defined in the model. This is important for working in a Linked Data context, where you will want access to data you may not have known about when your models were written.

```ruby
related = Thing.new

related << RDF::Statement(related, RDF::Vocab::DC.relation, obj)
related << RDF::Statement(related, RDF::Vocab::DC.subject, 'ActiveTriples')

related.query(subject:   related,
              predicate: RDF::Vocab::DC.relation).each_statement do |s,p,o|
  puts o
end
# => http://example.org/things#123

related.query(subject:   related,
              predicate: RDF::Vocab::DC.subject).each_statement do |s,p,o|
  puts o
end
# => 'ActiveTriples'
```

Any operation you can run against an RDF::Graph works with RDFSources, too. Or you can use generic setters and getters with URI predicates:

```ruby
related.set_value(RDF::Vocab::DC.relation, obj)
related.set_value(RDF::Vocab::DC.subject,  'ActiveTriples')

related.get_values(RDF::Vocab::DC.relation)
# => [#<Thing:0x3f949c6a2294(default)>]

related.get_values(RDF::Vocab::DC.subject)
# => ["ActiveTriples"]
```

Some convienience methods provide support for handling data from web sources:
  * `fetch` loads data from the RDFSource's #rdf_subject URI
  * `rdf_label` queries across common (& configured) label fields and returning the best match

```ruby
require 'linkeddata' # to support various serializations

uri = 'http://dbpedia.org/resource/Oregon_State_University'

osu = ActiveTriples::Resource.new uri
osu.fetch

osu.rdf_label
# => ["Oregon State University", "Oregon State University", "Université d'État de l'Oregon", "Oregon State University", "Oregon State University", "オレゴン州立大学", "Universidad Estatal de Oregón", "Oregon State University", "俄勒岡州立大學", "Universidade do Estado do Oregon"]
```

Typed Data
-----------

Typed literals are handled natively through Ruby types and [RDF::Literal](https://github.com/ruby-rdf/rdf/tree/develop/lib/rdf/model/literal). There is no need to register a specific type for a property, simply pass the setter the appropriate typed data. See the examples in the RDF::Literal documentation for futher information about supported datatypes.

```ruby
Thing.property :date, predicate: RDF::Vocab::DC.date

my_thing      = Thing.new
my_thing.date = Date.today

puts my_thing.dump :ntriples
# _:g70072864570340 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .
# _:g70072864570340 <http://purl.org/dc/terms/date> "2014-06-19Z"^^<http://www.w3.org/2001/XMLSchema#date> .
```

Data is cast back to the appropriate class when it is accessed.

```ruby
my_thing.date
# => [Thu, 19 Jun 2014]
```

Note that you can mix types on a single property.

```ruby
my_thing.date << DateTime.now
my_thing.date << "circa 2014"
my_thing.date
# => [Thu, 19 Jun 2014, Thu, 19 Jun 2014 11:39:21 -0700, "circa 2014"]

puts my_thing.dump :ntriples
# _:g70072864570340 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .
# _:g70072864570340 <http://purl.org/dc/terms/date> "2014-06-19Z"^^<http://www.w3.org/2001/XMLSchema#date> .
# _:g70072864570340 <http://purl.org/dc/terms/date> "2014-06-19T11:39:21-07:00"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
# _:g70072864570340 <http://purl.org/dc/terms/date> "circa 2014" .
```

Repositories and Persistence
-----------------------------

Resources can persist to various databases and triplestores though integration with [RDF::Repository](http://rubydoc.info/github/ruby-rdf/rdf/RDF/Repository).

```ruby
# Registers in-memory repositories. Other implementations of
# RDF::Repository support persistence to (e.g.) triplestores & NoSQL
# databases.
ActiveTriples::Repositories.add_repository :default, RDF::Repository.new
ActiveTriples::Repositories.add_repository :people,  RDF::Repository.new

class Person
  include  ActiveTriples::RDFSource
  
  configure type:       RDF::Vocab::FOAF.Person,
            base_uri:   'http://example.org/people#',
            repository: :people
  property :name, predicate: RDF::Vocab::FOAF.name
end

class Thing
  include  ActiveTriples::RDFSource

  configure type:       RDF::OWL.Thing,
            base_uri:   'http://example.org/things#',
            repository: :default
            
  property :title,       predicate: RDF::Vocab::DC.title
  property :description, predicate: RDF::Vocab::DC.description
  property :creator,     predicate: RDF::Vocab::DC.creator, class_name: 'Person'
end

t         = Thing.new('1')
t.title   = 'A Thing'
t.creator = Person.new('1')

t.persisted? # => false

ActiveTriples::Repositories.repositories[:default].dump :ntriples
# => ""

t.creator.first.name = 'Tove'
t.persist!

puts ActiveTriples::Repositories.repositories[:default].dump :ntriples
# <http://example.org/things#1> <http://purl.org/dc/terms/title> "A Thing" .
# <http://example.org/things#1> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .
# <http://example.org/things#1> <http://purl.org/dc/terms/creator> <http://example.org/people#1> .
# <http://example.org/people#1> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> .
# <http://example.org/people#1> <http://xmlns.com/foaf/0.1/name> "Tove" .
```

Contributing
-------------

Please observe the following guidelines:

 - Do your work in a feature branch based on ```develop``` and rebase before submitting a pull request.
 - Write tests for your contributions.
 - Document every method you add using YARD annotations. (_Note: Annotations are sparse in the existing codebase, help us fix that!_)
 - Organize your commits into logical units.
 - Don't leave trailing whitespace (i.e. run ```git diff --check``` before committing).
 - Use [well formed](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) commit messages.


CHANGES.md
1.1.0
----
  - Add support for RDF.rb 3.0

1.0.0
----
  - Finalize 1.0.0 APIs

0.11.0
----
  - Reworks ParentStrategy to use an Transaction over a well defined
  "extended bounded description".
  - Converts Relation to an Enumerable for more efficient access to
  property values.
  - Cleans up .Relation equality and added `#|` & `#&` for Set style
  comparison.
  - Deprecates `Relation#first_or_create`.
  - Removes dependency on the `linkeddata` gem. Users should require
  individual RDF libraries as needed.
  - Adds inheritance of configured types when subclassing an
  `RDFSource`.
  - Uses `URI#intern` to avoid repeated allocations of common URIs.
  - Changes handling of language and datatyped Literals in
  `Relation#each` & `#to_a`;
    - now returns `RDF::Literal` for lanugage tagged strings and
    for unknown datatypes.

0.10.0
----
  - Fix Identifiable for ActiveFedora [Trey Pendragon]
  - Upgrade to RDF.rb 2.0.2 [Tom Johnson]
  - Avoid touching each_statement on initialization [Tom Johnson]
  - Drop support for Ruby 2.0

0.9.0
-----
 - Upgrade to RDF.rb 2.0

0.8.3
-----
 - Do not include all predicates from other subjects in unregistered_predicates
 - Minor corrections to README.md

0.8.2
-----
* Allow PersistenceStrategy set by property config [Tom Johnson]
* Extend NodeConfig for arbitrary properties [Tom Johnson]
* Add `Relation#delete?` and `#swap` [Tom Johnson]
* Make Relation#delete singular; add #subtract [Tom Johnson]
* Re-add `Relation#delete` with a new implementation [Tom Johnson]
* Documentation and formatting cleanup on Relation [Tom Johnson]
* Remove Relation#reset! [Tom Johnson]
* Remove `Relation#[]=` [Tom Johnson]
* Remove Relation#delete [Tom Johnson]
* FIX #200 Use == instead of eql? for resource equality test [E. Lynette Rayle]
* Add some docs and tests for `PropertyBuilder` [Tom Johnson]
* Test undefined property on Relation#set [Tom Johnson]
* add loaded flag to lazy load property sources with parent_strategy [E. Lynette Rayle]
* Add deprecation warning [MJ Suhonos]
* Rename #obj to #source for clarity [MJ Suhonos]
* Update Guardfile [MJ Suhonos]
* Update comment terminology [MJ Suhonos]
* Remove reference to concrete persistable from abstract class [MJ Suhonos]
* Allow fetch to pass args to RDF::Reader.open [Justin Coyne]
* Delegate join to Relation. [Trey Terrell]
* Change ParentStrategy usage [Tom Johnson]
* Delegate #size in Relation. [Trey Terrell]
* Remove singleton_class call from RepositoryStrategy [Tom Johnson]
* Convert ancestors enumerator method to Class [Tom Johnson]
* Avoid circularity in `ParentStrategy` [Tom Johnson]
* Update .travis.yml Rubies to match RDF.rb [Tom Johnson]
* Add frozen_string_literals pragma [Justin Coyne]
* Fixup specs [Tom Johnson]
* Specs/docs for ValueError scenarios on #set_value [Tom Johnson]
* Add initial docs/specs for `RDFSource#attributes` [Tom Johnson]
* Add docs and specs for `Relation#first_or_create` [Tom Johnson]
* Add tests & docs for `Relation#build` [Tom Johnson]
* Make `Relation#clear` atomic [Tom Johnson]
* Test `RDFSource#rdf_label` [Tom Johnson]
* Fix for circular parent relationships [Tom Johnson]
* Add documentation and tests for key Relation [Tom Johnson]
* Refactor Relation#value_arguments [Tom Johnson]
* Move property methods from RDFSource in Properties [Tom Johnson]
* Add documentation and some tests for Properties [Tom Johnson]
* Refactor and rearrange RDFSource & Reflection [Tom Johnson]
* Finish documentation and testing of Reflection [Tom Johnson]
* Cleaner handling of undefined properties [Tom Johnson]
* Add some tests for `#get_values` [Tom Johnson]
* Make `#set_value` return the Relation it updates [Tom Johnson]
* Add triple in `#set_value` when argument is self [Tom Johnson]
* Removes odd logic surrounding property clearance [Tom Johnson]
* Removing OpenStruct in favor of PORO [Jeremy Friesen]
* Switching from Array() to Array.wrap [Jeremy Friesen]
* Fixup RDF.rb and rdf-spec dependencies [Tom Johnson]
* Track latest `rdf-spec` and `rdf-vocab` [Tom Johnson]
* Complete update to use RDF::Vocab [Tom Johnson]
* Use rdf-vocab for vocabularies [Justin Coyne]
* Add `RDFSource#graph_name` since quads aren't returned [Tom Johnson]
* Update specs to use a managable invalid statement [Tom Johnson]
* Test with develop version of rdf-spec [Justin Coyne]
* Graph#query is faster than using Queryable#query [Justin Coyne]
* Update error handling specs. [Tom Johnson]
* Adds a default block to `RDFSource#fetch` [Tom Johnson]
* Run `ActiveModel` lints on `RDFSource` [Tom Johnson]
* Fix ActiveModel linter to test `#to_key` [Tom Johnson]
* Remove unnecessary lines in `parent_strategy_spec` [Tom Johnson]
* Make `Relation#predicate` a public method [Tom Johnson]
* Cleanup and unit tests for `RDFSource#get_value` [Tom Johnson]
* Test ActiveModel validations with RDF's `#valid?` [Tom Johnson]
* Children should not be persisted? unless their parents are. Fixes #148 [Justin Coyne]


0.8.1
-----

  - Reverts changing `Relation`'s delete methods to remove all values until we
  have a clear path forward for those depending on that functionality.

0.8.0
-----
  - Adds RDF.rb interfaces to `RDFSource`, improving interoperability
  with other `ruby-rdf` packages.
  - Introduces a defined `Persistable` interface and
  `PersistenceStrategies`.
  - Changes `Relation`'s delete methods to remove all values, instead of
  trying to maintain a predicate -> class pair on the property
  definitions in some cases. The previous functionality was unclear and
  unreliable.
  - Adds a `Schema` concept, for defining property definitions that are
   portable across `RDFSource` types.


0.7.0
-----

__ATTN: This release withdraws support for Ruby 1.9__

  - Removes `#solrize` which was a badly named holdever from the
  ActiveFedroa days.
  - Fixes a bug on properties defined without a predicate. They are now
  rejected.
  - Disallows setting properties on the `ActiveTriples::Resource` base

lib/active_triples.rb
require 'rdf'
require 'active_triples/version'
require 'active_support'

##
# An ActiveModel compliant ObjectGraphMapper for RDF data. 
# 
# Models graphs as `RDFSources` with property/attribute configuration, 
# accessors, and other methods to support Linked Data in a Ruby enviornment. 
#
# @example modeling a simple resource
#   class Thing
#     include  ActiveTriples::RDFSource
#     configure :type => RDF::OWL.Thing, :base_uri => 'http://example.org/things#'
#     property :title, :predicate => RDF::DC.title
#     property :description, :predicate => RDF::DC.description
#   end
#
#   obj = Thing.new('123')
#   obj.title = 'Resource'
#   obj.description = 'A resource.'
#   obj.dump :ntriples 
#    # => "<http://example.org/things#123> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2002/07/owl#Thing> .\n<http://example.org/things#123> <http://purl.org/dc/terms/title> \"Resource\" .\n<http://example.org/things#123> <http://purl.org/dc/terms/description> \"A resource.\" .\n"
#
# @see http://www.w3.org/TR/2014/REC-rdf11-concepts-20140225/#change-over-time 
#   RDF Concepts and Abstract Syntax for an informal definition of an RDF 
#   Source.
module ActiveTriples
  extend ActiveSupport::Autoload
  eager_autoload do
    autoload :RDFSource
    autoload :Resource
    autoload :List
    autoload :Relation
    autoload :Configurable
    autoload :Persistable
    autoload :Properties
    autoload :PropertyBuilder
    autoload :Reflection
    autoload :Repositories
    autoload :NodeConfig
    autoload :NestedAttributes
    autoload :Identifiable
    autoload :Configuration
    autoload :Schema
    autoload :Property
    autoload :ExtensionStrategy

    # persistence strategies
    autoload :PersistenceStrategy,
             'active_triples/persistence_strategies/persistence_strategy'
    autoload :ParentStrategy,
             'active_triples/persistence_strategies/parent_strategy'
    autoload :RepositoryStrategy,
             'active_triples/persistence_strategies/repository_strategy'

    # error classes
    autoload :UndefinedPropertyError
  end
  
  ##
  # Raised when a declared repository doesn't have a definition
  class RepositoryNotFoundError < StandardError
  end

  ##
  # Converts a string for a class or module into a a constant. This will find
  # classes in or above a given container class.
  #
  # @example finding a class in Kernal
  #    ActiveTriples.class_from_string('MyClass') # => MyClass
  #
  # @example finding a class in a module
  #    ActiveTriples.class_from_string('MyClass', MyModule) 
  #    # => MyModule::MyClass
  #
  # @example when a class exists above the module, but not in it
  #    ActiveTriples.class_from_string('Object', MyModule) 
  #    # => Object
  #
  # @param class_name [String]
  # @param container_class
  #
  # @return [Class]
  def self.class_from_string(class_name, container_class=Kernel)
    container_class = container_class.name if container_class.is_a? Module
    container_parts = container_class.split('::')
    (container_parts + class_name.split('::'))
      .flatten.inject(Kernel) do |mod, class_name|
      if mod == Kernel
        Object.const_get(class_name)
      elsif mod.const_defined? class_name.to_sym
        mod.const_get(class_name)
      else
        container_parts.pop
        class_from_string(class_name, container_parts.join('::'))
      end
    end
  end

  ##
  # A simplified, Belgian version of this software
  def self.ActiveTripels
    puts <<-eos

        ###########
        ******o****
         **o******
          *******
           \\***/
            | |
            ( )
            / \\
        ,---------.

eos
"Yum"
  end
end
